extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs

```diff
@@ -135,72 +93,143 @@ private bool IsBlobExtension(ModelBindingData bindingData)
+                return await ToTargetTypeAsync(targetType, container, blobData.BlobName!);
+            }
         }
 
-        private async Task<object?> ToTargetTypeAsync(Type targetType, string connectionName, string containerName, string blobName) => targetType switch
-        {
-            Type _ when targetType == typeof(string) => await GetBlobStringAsync(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(Stream) => await GetBlobStreamAsync(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(byte[]) => await GetBlobBinaryDataAsync(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(BlobBaseClient) => CreateBlobClient<BlobBaseClient>(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(BlobClient) => CreateBlobClient<BlobClient>(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(BlockBlobClient) => CreateBlobClient<BlockBlobClient>(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(PageBlobClient) => CreateBlobClient<PageBlobClient>(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(AppendBlobClient) => CreateBlobClient<AppendBlobClient>(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(BlobContainerClient) => CreateBlobContainerClient(connectionName, containerName),
-            _ => await DeserializeToTargetObjectAsync(targetType, connectionName, containerName, blobName)
+        private bool IsCollectionBinding(Type type, string blobName)
+        {
+            bool isContainer = string.IsNullOrEmpty(blobName);
+
+            if (type == typeof(string) || type == typeof(byte[]))
+            {
+                return false;
+            }
+
+            bool isCollectionType = type.IsArray || typeof(IEnumerable).IsAssignableFrom(type);
+
+            if (!isCollectionType)
+            {
+                return false;
+            }
+
+            if (!isContainer)
+            {
+                throw new InvalidOperationException("Collections are not supported when binding to a specific blob file.");
+            }
+
+            return true;
+        }
+
+        private async Task<object> BindToCollectionAsync(Type targetType, BlobContainerClient container)
+        {
+            Type elementType = targetType.IsArray ? targetType.GetElementType() : targetType.GenericTypeArguments[0];
+
+            if (elementType == typeof(BlobContainerClient))
+            {
+                throw new InvalidOperationException("Collections of BlobContainerClient are not supported.");
+            }
+
+            var blobCount = 0;
+            var resultType = typeof(List<>).MakeGenericType(elementType);
+            var result = (IList)Activator.CreateInstance(resultType);
```

Comment:
@fabiocav so this used to be `IList result = Array.CreateInstance(elementType, collectionModelBindingData.ModelBindingDataArray.Length);`

But we don't know the length here and so I switched this around to creating a list and later converting to an array - is there a better way to do this?

Reply:
I can't think of anything better given the constraints.

Reply:
Alternatives I'm thinking of would unnecessarily increase complexity here for very little benefit. Let's keep this.